mlir/lib/Dialect/LLVMIR/IR/LLVMAttrs.cpp

```diff
@@ -109,3 +119,251 @@ bool MemoryEffectsAttr::isReadWrite() {
+// DIExpression
+//===----------------------------------------------------------------------===//
+
+DIExpressionAttr DIExpressionAttr::get(MLIRContext *context) {
+  return get(context, ArrayRef<uint64_t>({}));
+}
+
+#define _FORMAT_DW_OP(x) #x
+#define FORMAT_DW_OP(NAME, VENDOR) _FORMAT_DW_OP(DW_OP_##VENDOR##_##NAME)
+
+size_t queryNumOpParams(llvm::dwarf::LocationAtom op) {
+  switch (op) {
+  case llvm::dwarf::DW_OP_addr:
+  case llvm::dwarf::DW_OP_const1u:
+  case llvm::dwarf::DW_OP_const1s:
+  case llvm::dwarf::DW_OP_const2u:
+  case llvm::dwarf::DW_OP_const2s:
+  case llvm::dwarf::DW_OP_const4u:
+  case llvm::dwarf::DW_OP_const4s:
+  case llvm::dwarf::DW_OP_const8u:
+  case llvm::dwarf::DW_OP_const8s:
+  case llvm::dwarf::DW_OP_constu:
+  case llvm::dwarf::DW_OP_consts:
+  case llvm::dwarf::DW_OP_pick:
+  case llvm::dwarf::DW_OP_plus_uconst:
+  case llvm::dwarf::DW_OP_bra:
+  case llvm::dwarf::DW_OP_skip:
+  case llvm::dwarf::DW_OP_breg0:
+  case llvm::dwarf::DW_OP_breg1:
+  case llvm::dwarf::DW_OP_breg2:
+  case llvm::dwarf::DW_OP_breg3:
+  case llvm::dwarf::DW_OP_breg4:
+  case llvm::dwarf::DW_OP_breg5:
+  case llvm::dwarf::DW_OP_breg6:
+  case llvm::dwarf::DW_OP_breg7:
+  case llvm::dwarf::DW_OP_breg8:
+  case llvm::dwarf::DW_OP_breg9:
+  case llvm::dwarf::DW_OP_breg10:
+  case llvm::dwarf::DW_OP_breg11:
+  case llvm::dwarf::DW_OP_breg12:
+  case llvm::dwarf::DW_OP_breg13:
+  case llvm::dwarf::DW_OP_breg14:
+  case llvm::dwarf::DW_OP_breg15:
+  case llvm::dwarf::DW_OP_breg16:
+  case llvm::dwarf::DW_OP_breg17:
+  case llvm::dwarf::DW_OP_breg18:
+  case llvm::dwarf::DW_OP_breg19:
+  case llvm::dwarf::DW_OP_breg20:
+  case llvm::dwarf::DW_OP_breg21:
+  case llvm::dwarf::DW_OP_breg22:
+  case llvm::dwarf::DW_OP_breg23:
+  case llvm::dwarf::DW_OP_breg24:
+  case llvm::dwarf::DW_OP_breg25:
+  case llvm::dwarf::DW_OP_breg26:
+  case llvm::dwarf::DW_OP_breg27:
+  case llvm::dwarf::DW_OP_breg28:
+  case llvm::dwarf::DW_OP_breg29:
+  case llvm::dwarf::DW_OP_breg30:
+  case llvm::dwarf::DW_OP_breg31:
+  case llvm::dwarf::DW_OP_regx:
+  case llvm::dwarf::DW_OP_fbreg:
+  case llvm::dwarf::DW_OP_piece:
+  case llvm::dwarf::DW_OP_deref_size:
+  case llvm::dwarf::DW_OP_xderef_size:
+  case llvm::dwarf::DW_OP_call2:
+  case llvm::dwarf::DW_OP_call4:
+  case llvm::dwarf::DW_OP_call_ref:
+  case llvm::dwarf::DW_OP_addrx:
+  case llvm::dwarf::DW_OP_constx:
+  case llvm::dwarf::DW_OP_convert:
+  case llvm::dwarf::DW_OP_reinterpret:
+    return 1;
+  case llvm::dwarf::DW_OP_bregx:
+  case llvm::dwarf::DW_OP_bit_piece:
+  case llvm::dwarf::DW_OP_implicit_value:
+  case llvm::dwarf::DW_OP_implicit_pointer:
+  case llvm::dwarf::DW_OP_entry_value:
+  case llvm::dwarf::DW_OP_regval_type:
+  case llvm::dwarf::DW_OP_deref_type:
+  case llvm::dwarf::DW_OP_xderef_type:
+    return 2;
+  case llvm::dwarf::DW_OP_const_type:
+    return 3;
+  default:
+    return 0;
+  }
+}
+
+::mlir::Attribute DIExpressionAttr::parse(AsmParser &parser, Type type) {
```

Comment:
I wonder if there is a way to simplify / automate the printing and parsing a little bit. One established way of printing enum parameters is LLVM_DIParameter. It is used to print an parse enums inside DI attributes. LLVMAttrDefs.td contains some examples. It seems like this is not completely sufficient in your case though since there can be integer arguments. One solution to solve this could be to introduce a struct attribute that contains the enum as first argument and then a optional integer list as second argument. An example, that could serve as an inspiration is LLVM_DISubroutineTypeAttr.

It may be that I do not fully understand the requirements though. However, I think it would be great to try to express things mostly in tablegen if possible.

Reply:
I can go through the trouble of defining the expression operators in tablegen as it would be a better long term approach, but there would be close to 200 of them. The preprocessor define approach nicely automated this and stays consistent with LLVM automatically, but I think it introduces overhead during parsing/printing since the switch may actually touch the whole range of them. I can take the declarative approach if you think it is worthwhile for this.

Reply:
I think I agree with the goal of being in sync with LLVM and defining 200 different attributes does not sound like a good idea! 

From the examples in the PR, I got the understanding that a DIExpression is a list of opcodes (that have a name which we should have in an enum or similar) and of integers. So the primary problem to solve is how we want to pretty print the opcodes. Do you think LLVM_DIParameter or a slightly modified class could work for this? I guess the question is if LLVM provides a method that can be used to parse and print the opcode.

Once we have a way to express an opcode inside an attribute, we could define one attribute:

```
def LLVM_DIExpressionElemAttr : LLVM_Attr<"DIExpressionElem", "di_expression_elem"> {
  let parameters = (ins
    LLVM_DIExpresionOpcode:$opcode,
    OptionalArrayRefParameter<"int64_t">:$arguments
  );
```
Ah DIExpression would then be a list of such DIExpressionElemAttr.

Could something like this make sense? Or are the integers not really bound to the opcode before? 

I assume that this complicates the import and export a little bit since the underlying structure does not exactly match.

Reply:
This can work and would help simplify the parsing. I'll implement it

Reply:
This worked quite well, but there is a gap in functionality for expressions like this:
`DW_OP_LLVM_convert, 16, DW_ATE_signed`

I think I can change DIExpressionElemAttr to use an ArrayAttr so int operands can be mixed with encoding keywords.

Reply:
Yeah I guess then it needs to be an ArrayAttr. I start to wonder if the LLVM_DIExpressionElemAttr still makes sense? Or should the LLVM_DIExpressionAttr contain an ArrayAttr that can contain opcodes encodings and integers? I think logically it would be nice to group the opcode with its operands. However, being closer to LLVM may make the import and export easier. In any case I like that we can use LLVM_DIParameter 👍

Reply:
Alternative may be to have an optional encoding field in DIExpressionElemAttr (assuming the order does not matter). The question is a bit how many more enums there are...

Reply:
LLVM_DIExpressionElemAttr still makes sense to me as my usage of it just describes an operation (I might rename this as such). I think I just need another attribute to hold an operand value and info about it that handles parsing and printing. I suppose this new attribute could take a list of LLVM_DIParameters and builds the parsing and printing logic based on that. During parsing it could either parse an integer directly or if a keyword is parsed, it can test it against the DIParameters. If one satisfied it, I could store some identifier that would be referenced when printing the value out later.

...in theory 😵‍💫

Reply:
I'm almost positive I won't be able to get this to work using the existing stuff implemented in tablegen. In order to do this properly, I would need to somehow store both the value and what DWARF kind that value is together. THEN in order to support printing the textual form, you'd need to look at the kind and choose the correct printer.

Importing LLVM IR is a whole different can of worms. Frankly DIExpression only will give a list of ints to make sense of. Figuring out operations was easy and I already implemented that, but now figuring out which operand maps to which DWARF kind is a rabbit hole I currently want to climb out of 🕳️ .

I think for the sake of getting this out, I'd rather not spend more time on getting expressions 1-to-1 with LLVM's representation. Operands as ints will work for now and anyone reading/handwriting IR can refer to the DWARF spec in the meantime. Otherwise, LLVM has enums for all of this stuff that can be used in code.

Reply:
``> I'm almost positive I won't be able to get this to work using the existing stuff implemented in tablegen. In order to do this properly, I would need to somehow store both the value and what DWARF kind that value is together. THEN in order to support printing the textual form, you'd need to look at the kind and choose the correct printer.

Yes there is no good support for such a sum type I think. The best that comes to my mind for storing one value in the DIExpression array could be to have an attribute that contains optional fields for the enums we support and an optional integer field that is used for values that do not map to any of the enums:
```
def LLVM_DIExpressionElemAttr : LLVM_Attr<"DIExpressionElem", "di_expression_elem"> {
  let parameters = (ins 
    OptionalParameter<LLVM_DIOpcodeParameter>:$opcode,
    OptionalParameter<LLVM_DIEncodingParameter>:$encoding,
    OptionalParameter<"int64_t">:$value
  );
}
```
Did not try this but in principle it should work. Also note that this would be a flat representation that does not group the integers that belong together. The import / export of this should be fairly easy since we can use the existing LLVM methods to check if an integer value maps to a specific enum.

Another approach could be to add a custom printer and parser to the existing DIExpression that during printing / parsing checks for every integer if there is a string representation (again using the LLVM functions that are used in the printers and parsers of LLVM_DIParameter. I think this is the lowest complexity solution which stays close to LLVM and the printer / parser should have acceptable complexity. It is clearly less typed than the "sum type" approach.

Is the grouping you implemented in queryNumOpParams necessary? Isn't there an LLVM function we can query to get the number of parameters of an opcode? Overall it seems like this adds quite a bit of complexity and maybe an easy approach could be to just use a flat list for the DIExpression?

 > I think for the sake of getting this out, I'd rather not spend more time on getting expressions 1-to-1 with LLVM's representation.

Also feel free to use DIExpression more or less as is yes. This can also be improved in a separate PR later on and you can use this PR to introduce the GlobalVariableExpression.

Reply:
Actually, I'll give it another try.

```
static void writeDIExpression(raw_ostream &Out, const DIExpression *N,
                              AsmWriterContext &WriterCtx) {
  Out << "!DIExpression(";
  FieldSeparator FS;
  if (N->isValid()) {
    for (const DIExpression::ExprOperand &Op : N->expr_ops()) {
      auto OpStr = dwarf::OperationEncodingString(Op.getOp());
      assert(!OpStr.empty() && "Expected valid opcode");

      Out << FS << OpStr;
      if (Op.getOp() == dwarf::DW_OP_LLVM_convert) {
        Out << FS << Op.getArg(0);
        Out << FS << dwarf::AttributeEncodingString(Op.getArg(1));
      } else {
        for (unsigned A = 0, AE = Op.getNumArgs(); A != AE; ++A)
          Out << FS << Op.getArg(A);
      }
    }
  } else {
    for (const auto &I : N->getElements())
      Out << FS << I;
  }
  Out << ")";
}
```
LLVM only cares to print the textual form of a value specifically for DW_OP_LLVM_convert. I don't know why I didn't think to look here.

Reply:
> Is the grouping you implemented in queryNumOpParams necessary? Isn't there an LLVM function we can query to get the number of parameters of an opcode? Overall it seems like this adds quite a bit of complexity and maybe an easy approach could be to just use a flat list for the DIExpression?

Yes this is necessary if you want to parse operations in their textual form, since you would need to know when the next operation starts and I need to know it again when rebuilding the attribute from LLVM IR. LLVM has a much simpler form of this method that I can steal and I can use LLVM's facilities in the importer.

Reply:
`#llvm.di_expression<[ DW_OP_deref,  DW_OP_constu(3),  DW_OP_plus,  DW_OP_LLVM_convert(4, DW_ATE_signed)]>`

I was able to get it to print in this form. I also was able to remove the query method and use LLVM's facilities everywhere.